### PR TITLE
[SPARK-27540][MLlib] Add 'meanAveragePrecision_at_k' metric to RankingMetrics

### PR DESCRIPTION
**Pull request by qb-tarushg**
## What changes were proposed in this pull request?

Added method 'meanAveragePrecisionAt' k to RankingMetrics.

This branch is rebased with squashed commits from https://github.com/apache/spark/pull/24458

## How was this patch tested?

Added code in the existing test RankingMetricsSuite.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala [hunk 1]
```diff
@@ -71,25 +71,56 @@ class RankingMetrics[T: ClassTag](predictionAndLabels: RDD[(Array[T], Array[T])]
...
+   * Computes the average precision at first k ranking positions of all the queries.
+   * If a query has an empty ground truth set, the value will be zero and a log
+   * warning is generated.
+   *
+   * @param pred predicted ranking
+   * @param lab ground truth
+   * @param k use the top k predicted ranking, must be positive
+   * @return average precision at first k ranking positions
+   */
+  private def averagePrecision(pred: Array[T], lab: Array[T], k: Int): Double = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala]: You can pass `labSet` here from callers to avoid computing it twice

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala (hunk 1 above)
Comment (qb-tarushg) [reply mllib/src/main/scala/org/apache/spark/mllib/evaluation/RankingMetrics.scala]: Thanks! done.

**Comment by srowen** [comment]
Oops, small problem:
```
**********************************************************************
File "/home/jenkins/workspace/NewSparkPullRequestBuilder/python/pyspark/mllib/evaluation.py", line 380, in __main__.RankingMetrics
Failed example:
    metrics.meanAveragePrecisionAt(1)
Expected:
    0.333334...
Got:
    0.3333333333333333
**********************************************************************
```

**Comment by qb-tarushg** [comment]
Sorry my bad, issue fixed.

**Comment by qb-tarushg** [comment]
@srowen Can we start again the test build?

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by qb-tarushg** [comment]
@srowen Thank you 😊

**Pull request closed without merging**